### PR TITLE
Summarize errors in final InstallError

### PR DESCRIPTION
**Pull request by haampie**
Many Spack tests use `SpackCommand('install')()` to install packages.

I'm not a big fan of how `SpackCommand` is sprinkled around in all the tests, because it's arguably too high-level for unit-tests. It'd be much better if we could immediately invoke `InstallCommand(...)` skipping argument parsing. But now that we use it everywhere, it'd be good if the actual exception that comes out of it after a failed install is informative.

Currently the output is:

```
E           spack.installer.InstallError: Installation request failed.  Refer to reported errors for failing package(s).
```

without the reported errors being anywhere in stdout/stderr because they are returned by `SpackCommand`, and that's the command throwing the above error.

With this PR, you get:

```
E          Multiple packages failed to install:
             zlib due to SpackError: ProcessError: blab la returned error code 2
             'make' '-j16' 'non-exising-target'
             libbsd due to SpackError: ProcessError: something something
             'make' '-j16' 'non-exising-target'
```

The idea is to only show the short error message, not the one that includes the parsed build log.

### CODE REVIEW

File: lib/spack/spack/installer.py [hunk 1]
```diff
@@ -1634,9 +1639,15 @@ def install(self):
                 self._update_failed(task, True, exc)
                 spack.hooks.on_install_failure(task.request.pkg.spec)
 
+                # Collect short versions of exceptions
+                if isinstance(exc, SpackError):
+                    install_errors.append((pkg.name, SpackError(exc.message)))
```
Comment (haampie) [inline lib/spack/spack/installer.py:1644]: this is a bit hacky, but `str` defaults to showing the long message, which includes a potentially very long summary of the build logs.

File: lib/spack/spack/installer.py [hunk 2]
```diff
@@ -1650,15 +1661,6 @@ def install(self):
...
                                        .format(fail_fast_err, str(exc)))
 
-                # Terminate at this point if the single explicit spec has
-                # failed to install.
-                if single_explicit_spec and task.explicit:
-                    raise
-
-                # Track explicit spec id and error to summarize when done
-                if task.explicit:
-                    failed_explicits.append((pkg_id, str(exc)))
```
Comment (haampie) [inline lib/spack/spack/installer.py:1660]: This is no longer necessary if we just show all the failures in short

File: lib/spack/spack/installer.py [hunk 3]
```diff
@@ -1650,15 +1661,6 @@ def install(self):
                     raise InstallError('{0}: {1}'
                                        .format(fail_fast_err, str(exc)))
 
-                # Terminate at this point if the single explicit spec has
-                # failed to install.
-                if single_explicit_spec and task.explicit:
-                    raise
```
Comment (haampie) [inline lib/spack/spack/installer.py:1656]: This is covered at the end

File: lib/spack/spack/installer.py [hunk 4]
```diff
@@ -1634,9 +1639,15 @@ def install(self):
                 self._update_failed(task, True, exc)
                 spack.hooks.on_install_failure(task.request.pkg.spec)
 
+                # Collect short versions of exceptions
+                if isinstance(exc, SpackError):
+                    install_errors.append((pkg.name, SpackError(exc.message)))
+                else:
+                    install_errors.append((pkg.name, exc))
```
Comment (tldahlgren) [inline lib/spack/spack/installer.py:1646]: Possible simplification:

```suggestion
                # Collect short versions of Spack exceptions
                err = SpackError(exc.message) if isinstance(exc, SpackError) else exc
                install_errors.append((pkg.name, err))
```

**Review by tldahlgren — COMMENTED** [review]
There was a reason for distinguishing between failures of explicit -- specs explicitly provided on the command line or explicitly listed in the environment configuration -- versus dependency specs that, IIRC, related to verbosity.  A *lot* has since changed though.

Seems reasonable to collect and summarize more failures that were just warning before.

Please rebase as some of the removed code (e.g., `exists_errors`) is no longer in `develop`.

**Pull request closed without merging**